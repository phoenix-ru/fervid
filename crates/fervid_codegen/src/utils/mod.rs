// to_ident_or_str

use std::fmt::{Write, Error};

use fervid_core::{FervidAtom, StrOrExpr};
use swc_core::{common::Span, ecma::ast::{ComputedPropName, Ident, EsReserved, IdentName, PropName, Str}};

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_ident(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_ident(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}

pub fn to_camelcase(s: &str, buf: &mut impl Write) -> Result<(), Error> {
    for (idx, word) in s.split('-').enumerate() {
        if idx == 0 {
            buf.write_str(word)?;
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                buf.write_char(ch_component)?;
            }

            // Push the rest of the word
            buf.write_str(&word[ch.len_utf8()..])?;
        }
    }

    Ok(())
}
