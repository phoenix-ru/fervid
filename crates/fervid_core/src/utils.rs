use swc_core::{common::Span, ecma::ast::{ComputedPropName, EsReserved, Ident, IdentName, PropName, Str}};

use crate::{AttributeOrBinding, FervidAtom, StrOrExpr, VBindDirective};

/// Checks whether the attributes name is the same as `expected_name`
#[inline]
pub fn check_attribute_name(attr: &AttributeOrBinding, expected_name: &str) -> bool {
    matches!(attr,
        AttributeOrBinding::RegularAttribute { name, .. } |
        AttributeOrBinding::VBind(VBindDirective { argument: Some(StrOrExpr::Str(name)), .. })
        if name == expected_name
    )
}

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    is_valid_propname(s)
}

#[inline]
pub fn is_valid_propname(s: &str) -> bool {
    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_propname(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_propname(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}
