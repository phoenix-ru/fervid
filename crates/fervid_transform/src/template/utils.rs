use fervid_core::FervidAtom;
use swc_core::{common::DUMMY_SP, ecma::ast::{ArrowExpr, BindingIdent, BlockStmtOrExpr, Expr, Ident, Pat}};

/// `foo-bar-baz` -> `FooBarBaz`
#[inline]
pub(crate) fn to_pascal_case(raw: &str, out: &mut String) {
    for word in raw.split('-') {
        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// `foo-bar-baz` -> `fooBarBaz`
#[inline]
pub(crate) fn to_camel_case(raw: &str, out: &mut String) {
    for (idx, word) in raw.split('-').enumerate() {
        if idx == 0 {
            out.push_str(word);
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// Wraps `expr` to `$event => (expr)`
#[inline]
pub fn wrap_in_event_arrow(expr: Box<Expr>) -> Box<Expr> {
    let evt_param = Pat::Ident(BindingIdent {
        id: Ident {
            span: DUMMY_SP,
            ctxt: Default::default(),
            sym: FervidAtom::from("$event"),
            optional: false,
        },
        type_ann: None,
    });

    Box::new(Expr::Arrow(ArrowExpr {
        span: DUMMY_SP,
        ctxt: Default::default(),
        params: vec![evt_param],
        body: Box::new(BlockStmtOrExpr::Expr(expr)),
        is_async: false,
        is_generator: false,
        type_params: None,
        return_type: None,
    }))
}
