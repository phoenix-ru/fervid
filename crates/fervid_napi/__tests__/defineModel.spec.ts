import { BindingTypes } from '..'
import { assertCode, compile } from './utils'
import { describe, expect, test } from 'vitest'

describe('defineModel()', () => {
  test('basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      const modelValue = defineModel({ required: true })
      const c = defineModel('count')
      const toString = defineModel('toString', { type: Function })
      </script>
      `,
      { outputSetupBindings: true }
    )
    assertCode(content)
    expect(content).toMatch('props: {')
    expect(content).toMatch(`"modelValue": {
            required: true
        },`)
    expect(content).toMatch(`'count': {},`)
    expect(content).toMatch(`'toString': {
            type: Function
        }`)
    expect(content).toMatch(
      `emits: [
        "update:modelValue",
        "update:count",
        "update:toString"
    ],`,
    )
    expect(content).toMatch(
      `const modelValue = _useModel(__props, "modelValue")`,
    )
    expect(content).toMatch(`const c = _useModel(__props, 'count')`)
    expect(content).toMatch(`const toString = _useModel(__props, 'toString')`)
    expect(content).toMatch(`return {
            modelValue,
            c,
            toString
        }`)
    expect(content).not.toMatch('defineModel')

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.PROPS,
      c: BindingTypes.SETUP_REF,
      toString: BindingTypes.SETUP_REF,
    })
  })

  test('w/ defineProps and defineEmits', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps({ foo: String })
      defineEmits(['change'])
      const count = defineModel({ default: 0 })
      </script>
    `,
    { outputSetupBindings: true }
    )
    assertCode(content)
    // The official compiler uses `mergeModels` which is not smart.
    // Fervid goes 1 step further and merges objects during compilation.
    expect(content).toMatch(`props: {
        foo: String,
        "modelValue": {
            default: 0
        },
        "modelModifiers": {}
    }`)
    expect(content).toMatch(`const count = _useModel(__props, "modelValue")`)
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      count: BindingTypes.SETUP_REF,
      foo: BindingTypes.PROPS,
      modelValue: BindingTypes.PROPS,
    })
  })

  test('w/ array props', () => {
    const { content, bindings } = compile(
      `
      <script setup>
      defineProps(['foo', 'bar'])
      const count = defineModel('count')
      </script>
    `,
    { outputSetupBindings: true }
    )
    assertCode(content)
    // TODO Also merge arrays into objects (refer to `_mergeModels` implementation)
    expect(content).toMatch(`props: _mergeModels([
        'foo',
        'bar'
    ], {
        'count': {},
        "countModifiers": {}
    })`)
    expect(content).toMatch(`const count = _useModel(__props, 'count')`)
    expect(content).not.toMatch('defineModel')
    expect(bindings).toStrictEqual({
      foo: BindingTypes.PROPS,
      bar: BindingTypes.PROPS,
      count: BindingTypes.SETUP_REF,
    })
  })

  test('w/ types, basic usage', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean | string>()
      const count = defineModel<number>('count')
      const disabled = defineModel<number>('disabled', { required: false })
      const any = defineModel<any | boolean>('any')
      </script>
      `,
      { outputSetupBindings: true }
    )
    assertCode(content)
    expect(content).toMatch(`"modelValue": {
            type: [
                String,
                Boolean
            ]
        }`)
    expect(content).toMatch('"modelModifiers": {}')
    expect(content).toMatch(`'count': {
            type: Number
        }`)
    expect(content).toMatch(
      `'disabled': {
            required: false,
            type: Number
        }`,
    )
    expect(content).toMatch(`'any': {
            type: Boolean,
            skipCheck: true
        }`)
    expect(content).toMatch(
      `emits: [
        "update:modelValue",
        "update:count",
        "update:disabled",
        "update:any"
    ]`,
    )

    expect(content).toMatch(
      `const modelValue = _useModel<boolean | string>(__props, "modelValue")`,
    )
    expect(content).toMatch(`const count = _useModel<number>(__props, 'count')`)
    expect(content).toMatch(
      `const disabled = _useModel<number>(__props, 'disabled')`,
    )
    expect(content).toMatch(
      `const any = _useModel<any | boolean>(__props, 'any')`,
    )

    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      count: BindingTypes.SETUP_REF,
      disabled: BindingTypes.SETUP_REF,
      any: BindingTypes.SETUP_REF,
    })
  })

  test('w/ types, production mode', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean>()
      const fn = defineModel<() => void>('fn')
      const fnWithDefault = defineModel<() => void>('fnWithDefault', { default: () => null })
      const str = defineModel<string>('str')
      const optional = defineModel<string>('optional', { required: false })
      </script>
      `,
      { outputSetupBindings: true },
      { isProduction: true },
    )
    assertCode(content)
    expect(content).toMatch(`"modelValue": {
            type: Boolean
        }`)
    expect(content).toMatch(`'fn': {}`)
    expect(content).toMatch(
      `'fnWithDefault': {
            default: ()=>null,
            type: Function
        },`,
    )
    expect(content).toMatch(`'str': {}`)
    expect(content).toMatch(`'optional': {
            required: false
        }`)
    expect(content).toMatch(
      `emits: [
        "update:modelValue",
        "update:fn",
        "update:fnWithDefault",
        "update:str",
        "update:optional"
    ]`,
    )
    expect(content).toMatch(
      `const modelValue = _useModel<boolean>(__props, "modelValue")`,
    )
    expect(content).toMatch(`const fn = _useModel<() => void>(__props, 'fn')`)
    expect(content).toMatch(`const str = _useModel<string>(__props, 'str')`)
    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
      fn: BindingTypes.SETUP_REF,
      fnWithDefault: BindingTypes.SETUP_REF,
      str: BindingTypes.SETUP_REF,
      optional: BindingTypes.SETUP_REF,
    })
  })

  test('w/ types, production mode, boolean + multiple types', () => {
    const { content } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean | string | {}>()
      </script>
      `,
      undefined,
      { isProduction: true },
    )
    assertCode(content)
    expect(content).toMatch(`"modelValue": {
            type: [
                String,
                Boolean,
                Object
            ]
        }`)
  })

  test('w/ types, production mode, function + runtime opts + multiple types', () => {
    const { content } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<number | (() => number)>({ default: () => 1 })
      </script>
      `,
      undefined,
      { isProduction: true },
    )
    assertCode(content)
    expect(content).toMatch(
      `"modelValue": {
            default: ()=>1,
            type: [
                Number,
                Function
            ]
        }`,
    )
  })

  test('get / set transformers', () => {
    const { content } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel({
        get(v) { return v - 1 },
        set: (v) => { return v + 1 },
        required: true
      })
      </script>
      `,
    )
    assertCode(content)
    expect(content).toMatch(/"modelValue": {\s+required: true,?\s+}/m)
    expect(content).toMatch(
      `_useModel(__props, "modelValue", {
            get (v) {
                return v - 1;
            },
            set: (v)=>{
                return v + 1;
            }
        })`,
    )

    const { content: content2 } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel({
        default: 0,
        get(v) { return v - 1 },
        required: true,
        set: (v) => { return v + 1 },
      })
      </script>
      `,
    )
    assertCode(content2)
    expect(content2).toMatch(
      /"modelValue": {\s+default: 0,\s+required: true,?\s+}/m,
    )
    expect(content2).toMatch(
      `_useModel(__props, "modelValue", {
            get (v) {
                return v - 1;
            },
            set: (v)=>{
                return v + 1;
            }
        })`,
    )
  })

  // TODO Support props destructure
  // test('usage w/ props destructure', () => {
  //   const { content } = compile(
  //     `
  //     <script setup lang="ts">
  //     const { x } = defineProps<{ x: number }>()
  //     const modelValue = defineModel({
  //       set: (v) => { return v + x }
  //     })
  //     </script>
  //     `,
  //     { propsDestructure: true },
  //   )
  //   assertCode(content)
  //   expect(content).toMatch(`set: (v) => { return v + __props.x }`)
  // })

  test('w/ Boolean And Function types, production mode', () => {
    const { content, bindings } = compile(
      `
      <script setup lang="ts">
      const modelValue = defineModel<boolean | string>()
      </script>
      `,
      { outputSetupBindings: true },
      { isProduction: true },
    )
    assertCode(content)
    expect(content).toMatch(`"modelValue": {
            type: [
                String,
                Boolean
            ]
        }`)
    expect(content).toMatch(`emits: [
        "update:modelValue"
    ]`)
    expect(content).toMatch(
      `const modelValue = _useModel<boolean | string>(__props, "modelValue")`,
    )
    expect(bindings).toStrictEqual({
      modelValue: BindingTypes.SETUP_REF,
    })
  })

  test('error on duplicate model name: default', () => {
    const { errors } = compile(
      `
      <script setup lang="ts">
      const model1 = defineModel()
      const model2 = defineModel()
      </script>
      `
    )
    
    expect(errors.length).toBe(1)
    expect(errors[0].message).toMatch('DuplicateDefineModelName')
  })

  test('error on duplicate model name: user', () => {
    const { errors } = compile(
      `
      <script setup lang="ts">
      const model1 = defineModel('foo')
      const model2 = defineModel('foo')
      </script>
      `
    )
    
    expect(errors.length).toBe(1)
    expect(errors[0].message).toMatch('DuplicateDefineModelName')
  })
})
