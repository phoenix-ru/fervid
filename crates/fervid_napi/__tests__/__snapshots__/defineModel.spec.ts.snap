// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineModel() > basic usage 1`] = `
"import { useModel as _useModel } from "vue";
export default {
    emits: [
        "update:modelValue",
        "update:count",
        "update:toString"
    ],
    props: {
        "modelValue": {
            required: true
        },
        "modelModifiers": {},
        'count': {},
        "countModifiers": {},
        'toString': {
            type: Function
        },
        "toStringModifiers": {}
    },
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const modelValue = _useModel(__props, "modelValue");
        const c = _useModel(__props, 'count');
        const toString = _useModel(__props, 'toString');
        return {
            modelValue,
            c,
            toString
        };
    }
};
"
`;

exports[`defineModel() > get / set transformers 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue"
    ],
    props: {
        "modelValue": {
            required: true
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const modelValue = _useModel(__props, "modelValue", {
            get (v) {
                return v - 1;
            },
            set: (v)=>{
                return v + 1;
            }
        });
        return {
            modelValue
        };
    }
});
"
`;

exports[`defineModel() > get / set transformers 2`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue"
    ],
    props: {
        "modelValue": {
            default: 0,
            required: true
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const modelValue = _useModel(__props, "modelValue", {
            get (v) {
                return v - 1;
            },
            set: (v)=>{
                return v + 1;
            }
        });
        return {
            modelValue
        };
    }
});
"
`;

exports[`defineModel() > w/ Boolean And Function types, production mode 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue"
    ],
    props: {
        "modelValue": {
            type: [
                String,
                Boolean
            ]
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props) {
        const modelValue = _useModel<boolean | string>(__props, "modelValue");
    }
});
"
`;

exports[`defineModel() > w/ array props 1`] = `
"import { mergeModels as _mergeModels, useModel as _useModel } from "vue";
export default {
    emits: [
        "update:count"
    ],
    props: _mergeModels([
        'foo',
        'bar'
    ], {
        'count': {},
        "countModifiers": {}
    }),
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const count = _useModel(__props, 'count');
        return {
            count
        };
    }
};
"
`;

exports[`defineModel() > w/ defineProps and defineEmits 1`] = `
"import { useModel as _useModel } from "vue";
export default {
    emits: [
        'change',
        "update:modelValue"
    ],
    props: {
        foo: String,
        "modelValue": {
            default: 0
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const count = _useModel(__props, "modelValue");
        return {
            count
        };
    }
};
"
`;

exports[`defineModel() > w/ types, basic usage 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue",
        "update:count",
        "update:disabled",
        "update:any"
    ],
    props: {
        "modelValue": {
            type: [
                String,
                Boolean
            ]
        },
        "modelModifiers": {},
        'count': {
            type: Number
        },
        "countModifiers": {},
        'disabled': {
            required: false,
            type: Number
        },
        "disabledModifiers": {},
        'any': {
            type: Boolean,
            skipCheck: true
        },
        "anyModifiers": {}
    },
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const modelValue = _useModel<boolean | string>(__props, "modelValue");
        const count = _useModel<number>(__props, 'count');
        const disabled = _useModel<number>(__props, 'disabled');
        const any = _useModel<any | boolean>(__props, 'any');
        return {
            modelValue,
            count,
            disabled,
            any
        };
    }
});
"
`;

exports[`defineModel() > w/ types, production mode 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue",
        "update:fn",
        "update:fnWithDefault",
        "update:str",
        "update:optional"
    ],
    props: {
        "modelValue": {
            type: Boolean
        },
        "modelModifiers": {},
        'fn': {},
        "fnModifiers": {},
        'fnWithDefault': {
            default: ()=>null,
            type: Function
        },
        "fnWithDefaultModifiers": {},
        'str': {},
        "strModifiers": {},
        'optional': {
            required: false
        },
        "optionalModifiers": {}
    },
    __name: "anonymous",
    setup (__props) {
        const modelValue = _useModel<boolean>(__props, "modelValue");
        const fn = _useModel<() => void>(__props, 'fn');
        const fnWithDefault = _useModel<() => void>(__props, 'fnWithDefault');
        const str = _useModel<string>(__props, 'str');
        const optional = _useModel<string>(__props, 'optional');
    }
});
"
`;

exports[`defineModel() > w/ types, production mode, boolean + multiple types 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue"
    ],
    props: {
        "modelValue": {
            type: [
                String,
                Boolean,
                Object
            ]
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props) {
        const modelValue = _useModel<boolean | string | {
        }>(__props, "modelValue");
    }
});
"
`;

exports[`defineModel() > w/ types, production mode, function + runtime opts + multiple types 1`] = `
"import { defineComponent as _defineComponent, useModel as _useModel } from "vue";
export default _defineComponent({
    emits: [
        "update:modelValue"
    ],
    props: {
        "modelValue": {
            default: ()=>1,
            type: [
                Number,
                Function
            ]
        },
        "modelModifiers": {}
    },
    __name: "anonymous",
    setup (__props) {
        const modelValue = _useModel<number | (() => number)>(__props, "modelValue");
    }
});
"
`;
